openvdb/openvdb/TypeList.h

```diff
@@ -0,0 +1,531 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+/// @file TypeList.h
+///
+/// @brief  A TypeList provides a compile time sequence of heterogeneous types
+///   which can be accessed, transformed and executed over in various ways.
+///   It incorporates a subset of functionality similar to boost::mpl::vector
+///   however provides most of its content through using declarations rather
+///   than additional typed classes.
+
+#ifndef OPENVDB_TYPELIST_HAS_BEEN_INCLUDED
+#define OPENVDB_TYPELIST_HAS_BEEN_INCLUDED
+
+#include "version.h"
+
+#include <tuple>
+#include <type_traits>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+
+/// @cond OPENVDB_TYPES_INTERNAL
+
+template<typename... Ts> struct TypeList; // forward declaration
+
+namespace internal {
+
+// Implementation details of @c TypeList
+
+/// @brief Dummy struct, used as the return type from invalid or out-of-range
+///        @c TypeList queries.
+struct NullType {};
```

Comment:
While we're moving this about - could we rename the namespace from internal to something that's less likely to clash (typelist_internal?). 

Using `internal::NullType` is rather generic.

Reply:
I've included `TypeList.h` back into `Types.h` so as not to to break anyone's code so doesn't really matter that it's being moved - but I'm happy to make this change